velox/exec/tests/CustomJoinBridgeTest.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/exec/JoinBridge.h"
+#include "velox/exec/tests/utils/OperatorTestBase.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+
+#include <gmock/gmock.h>
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+
+class MockJoinBridge : public JoinBridge {
+ public:
+  bool isMockJoinBridge() {
+    return true;
+  }
+};
+
+class MockJoinNode : public core::PlanNode {
+ public:
+  MockJoinNode(const core::PlanNodeId& id) : PlanNode(id) {}
+
+  MOCK_METHOD(const RowTypePtr&, outputType, (), (const, override));
+  MOCK_METHOD(
+      const std::vector<std::shared_ptr<const PlanNode>>&,
+      sources,
+      (),
+      (const, override));
+  MOCK_METHOD(std::string_view, name, (), (const, override));
+  MOCK_METHOD(
+      void,
+      addDetails,
+      (std::stringstream & stream),
+      (const, override));
+};
+
+class CustomJoinBridgeTranslator : public Operator::PlanNodeTranslator {
+  std::unique_ptr<Operator> translate(
+      DriverCtx* ctx,
+      int32_t id,
+      const std::shared_ptr<const core::PlanNode>& node) {
+    return nullptr;
```

Comment:
It would be nice to write a test that illustrates how this extensibility feature should be used and at the same time verifies it works. Hence, this method should not return nullptr as custom join node must return all 3 pieces: an operator, an operator supplier and a bridge.

Reply:
Would you mind me to reuse the join logic from HashJoin or CrossJoin since we should only need verify the translator API works here?

Reply:
We can have very simple implementation for the test join operator. It doesn't even need to perform the actual join. We just want to demonstrate how build-side operator can pass information to probe-side operator and how this information can be used by the probe-side operator. For example, build side can simple count the number of input rows and pass it via the join bridge. The probe side can could use this information as a "limit", e.g. return up to specified number of rows of input unmodified. Let's say there are 10 rows on build side and 100 on probe. If the communication works properly, build side will pass 10 via the bridge to the probe side and probe side would emit only 10 rows. Would that work?

Reply:
Thanks for your advice. Will follow up it.

Reply:
Update the custom join example